### PR TITLE
Current book refactor

### PR DESCRIPTION
**Pull request by jrodonohue**
Refactored code and integrated scrapers

### CODE REVIEW

**Review by mtuthill — CHANGES_REQUESTED** [review]
Great job with everything! You write some really clean code. The response time for the scraping is awesome and quick. Some overall comments:

- [ ] It doesn't seem that the previous button is still working from the final review/rating listing page to go back to the main page
- [ ] I tested a couple of search by title queries but the glass hotel information is showing up, not the book I searched for

**Comment by RipleyRyan** [comment]
I don't have much other to comment on, it seems like Marc covered all the important things. Nice work on this though Jack it looks very solid. Excited to get this merged up.

File: app/src/main/java/com/example/byitscover/scrapers/BarnesAndNobleScraper.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        Element reviewValue = (Element) document.getElementById("overviewSection").childNode(1).childNode(1)
+                .childNode(3).childNode(1).childNode(1).childNode(1);
+
+        Review review = new Review(null, Jsoup.clean(reviewValue.toString(), Whitelist.none()), null);
+        List<Review> reviews = new ArrayList<Review>();
+        reviews.add(review);
+
+        BookListing listing = new BookListing(new URL(ScraperHelper.getTopResultUrl(results)),
+                ScraperConstants.BARNES_AND_NOBLE,
+                new Book(query.getTitle(), query.getAuthor(), null, null),
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/scrapers/BarnesAndNobleScraper.java:59]: I didn't find any extraction of the author or title of the search result in the Barnes and Noble scraper, so I just used the query title.

File: app/src/main/java/com/example/byitscover/scrapers/GoodreadsScraper.java [hunk 2]
```diff
@@ -31,47 +44,54 @@
...
+        Element coverUrlElement = (Element) bookDocument.selectFirst("div.bookCoverPrimary")
+                .childNode(1).childNode(0);
+        URL coverUrl = new URL(coverUrlElement.absUrl("src"));
 
+        List<Review> reviews = new ArrayList<Review>();
+        reviews.add(review);
+
+        BookListing listing = new BookListing(new URL(ScraperHelper.getTopResultUrl(results)),
+                ScraperConstants.GOODREADS,
+                new Book(query.getTitle(), query.getAuthor(), null, null),
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/scrapers/GoodreadsScraper.java:87]: I copied and pasted my Barnes and Noble code, and I overlooked this. I've fixed this, but I still need to commit.

File: app/src/test/java/com/example/byitscover/scrapers/GoodreadsScraperTest.java [hunk 3]
```diff
@@ -16,21 +16,23 @@
     private static final String REVIEW_TEXT = "From the award-winning author of Station Eleven, an exhilarating novel set at the glittering intersection of two seemingly disparate events–a massive Ponzi scheme collapse and the mysterious disappearance of a woman from a ship at sea.";
     private static final Double RATING_VALUE = 3.71;
 
+    /*
     @Before
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/test/java/com/example/byitscover/scrapers/GoodreadsScraperTest.java:20]: Removed the file from the branch

File: app/src/main/java/com/example/byitscover/ReviewPage.java [hunk 4]
```diff
@@ -31,6 +72,40 @@ public ReviewPage() {
...
+                    setBarnesAndNobleInfo(view, listing);
+                }
+            }
+
+            for (BookListing listing : listings) {
+                if (listing.getWebsite().equals(ScraperConstants.BARNES_AND_NOBLE)) {
+                    setBarnesAndNobleInfo(view, listing);
+                    break;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/ReviewPage.java:101]: Removed duplicate

File: app/src/main/java/com/example/byitscover/ReviewPage.java [hunk 5]
```diff
@@ -31,6 +72,40 @@ public ReviewPage() {
...
+        } catch (ExecutionException ex) {
+            throw (RuntimeException)ex.getCause();
+        } catch (CancellationException | InterruptedException ex) {
+            // Display
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/ReviewPage.java:105]: The code is required to handle these exceptions by Java, but these exceptions should never occur. I updated the code to throw an `AssertionError` if any of the exceptions occur.

File: app/src/main/java/com/example/byitscover/ReviewPage.java [hunk 6]
```diff
@@ -45,37 +120,54 @@ public View onCreateView(
...
-        setAuthorAndTitle(view, instance);
-        setGoodreadsInfo(view, instance);
+                        return aggregate.scrape(query);
+                    }
+                },
+        this::onScraperCompletion);
+
+        setAuthorAndTitle(view, goodreadsResult.getBook());
+        setGoodreadsInfo(view, goodreadsResult);
         setAverageRatingValue(view);
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/ReviewPage.java:148]: This is to initialize the book display with a default book, but I'm not sure why it uses goodreadsResult. The code has been fixed to use the default book.

File: app/src/main/java/com/example/byitscover/ReviewPage.java [hunk 7]
```diff
@@ -85,13 +177,13 @@ public View onCreateView(
...
 
-        view.findViewById(R.id.searchByTitleButton).setOnClickListener(new View.OnClickListener() {
+        /*view.findViewById(R.id.searchByTitleButton).setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View view) {
                 NavHostFragment.findNavController(ReviewPage.this)
                         .navigate(R.id.action_from_review_to_first);
             }
-        });
+        });*/
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/ReviewPage.java:186]: Removed!

File: app/src/main/java/com/example/byitscover/helpers/Isbn.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+ * For example, 9780262033848 and 0-262-03384-4 are different representations of the same ISBN.
+ * This class is an ISBN, and it provides facilities for handling representations of ISBNs.
+ * All instances of this class must be instantiated from valid ISBN representations.
+ * The rest of the codebase may compare ISBNs without concern for the underlying representations.
+ *
+ * @author Jack
+ * @version 1.0
+ */
+
+// TODO: Add unit tests
```
[the remark this answers is not in the document]
Comment (jrodonohue) [reply app/src/main/java/com/example/byitscover/helpers/Isbn.java:17]: I already added unit tests, and I forgot to remove this. Thanks for pointing it out!

File: app/src/main/java/com/example/byitscover/scrapers/BarnesAndNobleScraper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtuthill) [reply app/src/main/java/com/example/byitscover/scrapers/BarnesAndNobleScraper.java:59]: Yeah definitely open to discussion on how we want to deal with populating the title and author on the review page moving forward. I think it should come from one of the scrapers to act as a bit of the spell check, but I'm not sure how to choose which one we go with. Maybe when we implement the "choose one of these 3 books" pop-up thing, we can use whatever info is actually selected by the user. I just think if we leave it just using the query we'll have misspellings/missing capitalization/gibberish that is outputted from the OCR. So I think let's leave this as is but for now just use the Goodreads results to populate the title and author and investigate further next week.

[the remark this answers is not in the document]
Comment (mtuthill) [reply app/src/main/java/com/example/byitscover/scrapers/BarnesAndNobleScraper.java:59]: Just some things for all of us to think about

**Review by mtuthill — APPROVED** [review]

**Pull request merged**